library/std/src/sys/windows/mutex.rs

```diff
@@ -60,10 +64,11 @@ impl Mutex {
         match kind() {
             Kind::SRWLock => c::AcquireSRWLockExclusive(raw(self)),
             Kind::CriticalSection => {
-                let re = self.remutex();
-                (*re).lock();
-                if !self.flag_locked() {
-                    (*re).unlock();
+                let inner = &mut *self.inner();
```

Comment:
are you sure we should take a `&mut` reference here rather than a `&` reference? 

since this is called before `.lock()`, it becomes possible for multiple threads to hold the same `&mut Inner` reference which is unsound. 

plus using `&mut` everywhere means the `Cell<bool>` is useless 🙃.

Reply:
Hm, the code already did this for `ReentrantMutex`. For that type, the `lock` function takes a `&self` but unsafely takes a `&mut` to a member right away.

I'll take a closer look.

Reply:
Changed all mut references and pointers to const ones. Not sure why the old code used `*mut`.

I also updated ReentrantMutex to avoid `&mut`s by swapping the `UnsafeCell<MaybeUninit>` for a `MaybeUninit<UnsafeCell>` and using [`UnsafeCell::raw_get`](https://doc.rust-lang.org/core/cell/struct.UnsafeCell.html#method.raw_get).